pkg/auth/auth.go

```diff
@@ -39,58 +40,97 @@ func (e ErrNewCredentialsInvalid) Unwrap() error {
 // Will fail over to DOCKER_CONFIG if REGISTRY_AUTH_FILE environment is not set
+//
+// WARNINGS:
+//   - In almost all invocations, expect this function to return ""; so it can not be used
+//     for directly accessing the file.
+//   - Use this only for commands that _read_ credentials, not write them.
+//     The path may refer to github.com/containers auth.json, or to Docker config.json,
+//     and the distinction is lost; writing auth.json data to config.json may not be consumable by Docker,
+//     or it may overwrite and discard unrelated Docker configuration set by the user.
 func GetDefaultAuthFile() string {
+	// Keep this in sync with the default logic in systemContextWithOptions!
+
 	if authfile := os.Getenv("REGISTRY_AUTH_FILE"); authfile != "" {
 		return authfile
 	}
+	// This pre-existing behavior is not conceptually consistent:
+	// If users have a ~/.docker/config.json in the default path, and no environment variable
+	// set, we read auth.json first, falling back to config.json;
+	// but if DOCKER_CONFIG is set, we read only config.json in that path, and we don’t read auth.json at all.
 	if authEnv := os.Getenv("DOCKER_CONFIG"); authEnv != "" {
 		return filepath.Join(authEnv, "config.json")
 	}
 	return ""
 }
 
-// CheckAuthFile validates filepath given by --authfile
-// used by command has --authfile flag
-func CheckAuthFile(authfile string) error {
-	if authfile == "" {
+// CheckAuthFile validates a path option, failing if the option is set but the referenced file is not accessible.
+func CheckAuthFile(pathOption string) error {
+	if pathOption == "" {
 		return nil
 	}
-	if _, err := os.Stat(authfile); err != nil {
-		return fmt.Errorf("checking authfile: %w", err)
+	if _, err := os.Stat(pathOption); err != nil {
+		return fmt.Errorf("credential file is not accessible: %w", err)
 	}
 	return nil
 }
 
 // systemContextWithOptions returns a version of sys
-// updated with authFile and certDir values (if they are not "").
+// updated with authFile, dockerCompatAuthFile and certDir values (if they are not "").
 // NOTE: this is a shallow copy that can be used and updated, but may share
 // data with the original parameter.
-func systemContextWithOptions(sys *types.SystemContext, authFile, certDir string) *types.SystemContext {
+func systemContextWithOptions(sys *types.SystemContext, authFile, dockerCompatAuthFile, certDir string) (*types.SystemContext, error) {
 	if sys != nil {
 		sysCopy := *sys
 		sys = &sysCopy
 	} else {
 		sys = &types.SystemContext{}
 	}
 
-	if authFile != "" {
+	defaultDockerConfigPath := filepath.Join(homedir.Get(), ".docker", "config.json")
+	switch {
+	case authFile != "" && dockerCompatAuthFile != "":
+		return nil, errors.New("--authfile and --compat-auth-file can not be set simultaneously")
+	case authFile != "":
+		if authFile == defaultDockerConfigPath {
+			logrus.Warn("--authfile points to ~/.docker/config.json, but the file format is not fully compatible; use --compat-auth-file instead")
```

Comment:
I am not sure how this package is used everywhere but pointing to cli options in error messages may not be optimal here. I see it used on the podman server API side (without any auth file option there so not a real issue but it might be an issue in the future). It is also used in the bindings test with an authfile so pointing to a cli option there would not make any sense.

Reply:
You’re completely right that conceptually, `auth.go` can be used without `cli.go`, so it should not assume any specific option names.

Can you suggest a practical option-name-independent wording? Especially in this message, my first stab is `credential file is set to ~/.docker/config.json, but the file format is not fully compatible; specify it using the )(Docker?) compatibility credential file option instead` , which is way too long, _and_ not all that helpful for users looking for the option name.

Looking at the callers, most users users of `pkg/auth` just call `GetDefaultAuthFile` and `CheckAuthFile`; the non-trivial callers are just the top-level CLI login/logout, except:
- Podman’s `pkg/api/handlers/compat/auth.go`. In that case neither `AuthFile` nor `CompatAuthFile` is specified, so these messages should not be reached.
  - (From a quick read without looking at the history, it seems to me that this API handler should not deal with auth files _at all_ (like `auth.Login` always reads them), and instead of the CLI login it should just call `docker.CheckAuth` directly; and `LoginOptions.NoWriteBack` could then be removed.)
- Podman’s `pkg/bindings/test/auth_test.go`. Awkward, but those errors should not happen with code as is, and future maintainers can hopefully deal with it.

So I think that referring to the `cli.go` option names directly is a a clearly imperfect but hopefully-acceptable compromise, but that’s just because I can’t think of good wording; I’ll be happy to change it to any other suggested text.

Reply:
Perhaps "saving credentials to ~/.docker/config.json but not using Docker-compatible credential format"?

Reply:
Thanks for that suggestion!

@Luap99 @mheon Texts updated; PTAL.